java-checks/src/main/java/org/sonar/java/checks/synchronization/DoubleCheckedLockingCheck.java

```diff
@@ -0,0 +1,233 @@
+
+  @Override
+  public void leaveNode(Tree tree) {
+    if (tree.is(IF_STATEMENT) && fieldNotInitializedTest(((IfStatementTree) tree).condition())) {
+      ifConditionStack.pollLast();
+    }
+    if (tree.is(SYNCHRONIZED_STATEMENT)) {
+      synchronizedStmtStack.pollLast();
+    }
+  }
+
+  private void visitIfStatement(IfStatementTree ifTree) {
+    if (!fieldNotInitializedTest(ifTree.condition())) {
+      return;
+    }
+    ifConditionStack.add(ifTree.condition());
+    if (!insideCriticalSection() || !isNestedIfStatement()) {
+      return;
+    }
+    patternDetected(ifTree);
+  }
+
+  private boolean isNestedIfStatement() {
+    return ifConditionStack.size() > 1;
+  }
+
+  /**
+   * {@code
+   * if (fieldInitializationTest) {
+   *   synchronized {
+   *     if (fieldInitializationTest) {
+   *       ...
+   *     }
+   *   }
+   * }
+   * }
+   *
+   * @param nestedIf - if inside synchronized section
+   */
+  private void patternDetected(IfStatementTree nestedIf) {
+    ExpressionTree parentIf = identicalConditionAlreadyOnStack(nestedIf.condition());
+    if (parentIf == null) {
+      return;
+    }
+    Symbol field = fieldFromEqCondition(nestedIf.condition());
+    if (field == null) {
+      return;
+    }
+    if (thenStmtInitializeField(nestedIf.thenStatement(), field) && !field.isVolatile() && !isImmutable(field)) {
+      reportIssue(parentIf.parent(), parentIf, MESSAGE);
+    }
+  }
+
+  /**
+   * This is naive, however can avoid the FP in the simplest cases
+   */
+  private static boolean isImmutable(Symbol field) {
+    Type fieldType = field.type();
+    if (!fieldType.isClass()) {
+      return false;
+    }
+    Collection<Symbol> members = field.type().symbol().memberSymbols();
+    Optional<Symbol> nonFinalField = members.stream()
+      .filter(symbol -> symbol.isVariableSymbol() && symbol.type().isPrimitive() && !symbol.isFinal())
+      .findAny();
+    return !nonFinalField.isPresent();
+  }
+
+  private static boolean thenStmtInitializeField(StatementTree statementTree, Symbol field) {
+    AssignmentVisitor visitor = new AssignmentVisitor(field);
+    statementTree.accept(visitor);
+    return visitor.assignmentToField;
+  }
+
+  private static Symbol fieldFromEqCondition(ExpressionTree condition) {
+    if (!condition.is(EQUAL_TO)) {
+      return null;
+    }
+    BinaryExpressionTree eqRelation = (BinaryExpressionTree) condition;
+    if (isField(eqRelation.leftOperand())) {
+      return symbolFromVariable(eqRelation.leftOperand());
+    }
+    if (isField(eqRelation.rightOperand())) {
+      return symbolFromVariable(eqRelation.rightOperand());
+    }
+    return null;
+  }
+
+  @CheckForNull
+  private ExpressionTree identicalConditionAlreadyOnStack(ExpressionTree nestedCondition) {
+
+    for (ExpressionTree parentCondition : ifConditionStack) {
+      Symbol symbol1 = fieldFromEqCondition(parentCondition);
+      Symbol symbol2 = fieldFromEqCondition(nestedCondition);
+      if (symbol1 == symbol2) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)